temporalio/workflow.py

```diff
@@ -746,6 +755,65 @@ def time_ns() -> int:
     return _Runtime.current().workflow_time_ns()
 
 
+def update(
+    fn: Optional[CallableSyncOrAsyncType] = None,
+    *,
+    name: Optional[str] = None,
+    dynamic: Optional[bool] = False,
+):
+    """Decorator for a workflow update handler method.
+
+    This is set on any async or non-async method that you wish to be called upon
+    receiving an update. If a function overrides one with this decorator, it too
+    must be decorated.
+
+    You may also optionally define a validator method that will be called before
+    this handler you have applied this decorator to. You can specify the validator
+    with ``@update_handler_function_name.validator``.
+
+    Update methods can only have positional parameters. Best practice for
+    non-dynamic update methods is to only take a single object/dataclass
+    argument that can accept more fields later if needed. The handler may return
+    a serializable value which will be sent back to the caller of the update.
+
+    Args:
+        fn: The function to decorate.
+        name: Update name. Defaults to method ``__name__``. Cannot be present
+            when ``dynamic`` is present.
+        dynamic: If true, this handles all updates not otherwise handled. The
+            parameters of the method must be self, a string name, and a
+            ``*args`` positional varargs. Cannot be present when ``name`` is
+            present.
+    """
+
+    def with_name(
+        name: Optional[str], fn: CallableSyncOrAsyncType
+    ) -> CallableSyncOrAsyncType:
+        defn = _UpdateDefinition(name=name, fn=fn, is_method=True)
+        if defn.dynamic_vararg:
+            raise RuntimeError(
+                "Dynamic updates do not support a vararg third param, use Sequence[RawValue]",
+            )
+        setattr(fn, "__temporal_update_definition", defn)
+        setattr(fn, "validator", partial(_update_validator, defn))
```

Comment:
Probably this should return a class rather than doing this, so IDE docstring/hints work on `validator`

Reply:
I think this is quite nice now

Reply:
This is still nicer I think, but avoids using a class decorator because of unlimited type hell.